tracer/src/Datadog.Trace/Debugger/Instrumentation/AsyncMethodDebuggerInvoker.SingleProbe.cs

```diff
@@ -21,10 +21,96 @@ namespace Datadog.Trace.Debugger.Instrumentation
     /// </summary>
     [Browsable(false)]
     [EditorBrowsable(EditorBrowsableState.Never)]
-    public static class AsyncMethodDebuggerInvoker
+    public static partial class AsyncMethodDebuggerInvoker
     {
         internal static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(AsyncMethodDebuggerInvoker));
 
+        /// <summary>
+        /// Determines if the instrumentation should call <see cref="UpdateProbeInfo{T}"/>.
+        /// </summary>
+        /// <param name="methodMetadataIndex">The unique index of the method.</param>
+        /// <param name="instrumentationSequence">The unique identifier of the instrumentation.</param>
+        /// <param name="state">If it the first entry to the state machine MoveNext method</param>
+        /// <returns>true if <see cref="UpdateProbeInfo{T}"/> should be called, false otherwise.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool ShouldUpdateProbeInfo(int methodMetadataIndex, int instrumentationSequence, ref AsyncDebuggerState state)
+        {
+            if (!MethodMetadataCollection.Instance.IndexExists(methodMetadataIndex))
+            {
+                return true;
+            }
+
+            ref var methodMetadataInfo = ref MethodMetadataCollection.Instance.Get(methodMetadataIndex);
+
+            var isDifferentSequence = methodMetadataInfo.InstrumentationSequence != instrumentationSequence;
+
+            if (IsReEnter(ref state) && isDifferentSequence)
+            {
+                // Re-enter but with different instrumentation version/sequence. Ditching state and avoiding the collection of further instrumentation data.
```

Comment:
Could be better not to use both `instrumentation version` and `instrumentation sequence` - instead to pick one (preferably the former) and use that across the entire PR

Reply:
I haven't resolved the naming just yet. Once I will, I'll work on aligning the naming everywhere.

Reply:
Done in d22e27f7934085a341580679dce764df01b41514.